### DISCUSSION TITLE
All tests being executed once after I opened my project (using vitest) in vscode, how can I disable this feature?

### DISCUSSION THREAD

**Discussion by Paper-Folding** [vitest-dev/vscode#398] [Q&A]
I'm also using playwright as my e2e test helper, every time project's being opened, the extension will execute all tests thus opening automated browser, which is annoying. Can I turn off this feature?
P.S. I'm not talking about enabling headless mode for e2e test.

**Comment by sheremet-va** [comment]
Do you mean, the tests use playwright CLI, or do you use Vitest to initiate playwright instance? Can you give any code examples? The extension shouldn't start any `playwright` instances during the collection phase unless you do it during collecting.

**Reply by Paper-Folding** [reply]
> Do you mean, the tests use playwright CLI, or do you use Vitest to initiate playwright instance? Can you give any code examples? The extension shouldn't start any `playwright` instances during the collection phase unless you do it during collecting.

Any test will be executed once when a vitest project opened in vscode with the extension installed. I confirmed it is the extension's fault because if I uninstall the extension the problem is gone.

**Comment by Paper-Folding** [comment]
Here is my vitest.config.js file (it'll start test site server once `npx vitest` is being executed, and `jsdom` is only for testing other helper functions):
```js
import { defineConfig } from "vite";
import { exec } from "child_process";

export default defineConfig(async () => {
    // start test site server, note that this is also being executed when project opened, which is undesirable behavior
    exec("yarn dev");

    return {
        test: {
            environment: "jsdom"
        }
    };
});
```

And here is one of my e2e test file (simplified):  
```ts
import { afterAll, beforeAll, describe, test} from "vitest";
import { Browser, Page, chromium } from "@playwright/test";

describe("sample e2e tests", () => {
    let browser: Browser;
    let page: Page;

    beforeAll(async () => {
        browser = await chromium.launch({ headless: false, args: ["--start-maximized"] });
        page = await browser.newPage({
            viewport: null
        });
        await page.goto("http://localhost:5173");  // goto test site
    });

    afterAll(async () => {
        await browser.close();
    });

    const testSteps = async () => {
        await page.waitForTimeout(2000);
        const cards = await page.locator(".cards").all();
        const bx0 = await cards[0].boundingBox();
        const bx1 = await cards[1].boundingBox();
        expect(Math.abs(bx0!.x + bx0!.width - bx1!.x)).toBeLessThan(1);
    };

    test("layout test", { timeout: 10000 }, async () => {
        await testSteps();
    });
});
```

Some of `package.json` scripts:
```
  "scripts": {
    "dev": "vite",
    "pretest": "playwright install chromium",
    "test": "vitest",
  },
```

**Reply by sheremet-va** [reply]
1. It is a very bad idea to spawn a process in the config definition. You can just start it in scripts directly: `vite & vitest`, Vitest needs to read it obviously.
2. It is a bug that Vitest calls `beforeAll`/`afterAll` if no tests are running

**Reply by Paper-Folding** [reply]
Okay, thanks for your reply. So Vitest extension will read config file on collection phase, right? And you mean vitest extension calls `beforeAll` / `afterAll` on collection phase, too ?

**Reply by Paper-Folding** [reply]
> 1. It is a very bad idea to spawn a process in the config definition. You can just start it in scripts directly: `vite & vitest`, Vitest needs to read it obviously.
> 2. It is a bug that Vitest calls `beforeAll`/`afterAll` if no tests are running

Acutually, `vite` will block the execution as it starts a server, so by writing `vite & vitest`, `vitest` is never gonna execute. That's reason why I put the `exec('npx vite')` inside `vitest.conf.js`
